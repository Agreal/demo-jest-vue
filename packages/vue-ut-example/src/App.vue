<template>
  <div>
    <h1>Latest Vue.js Commits</h1>
    <div class="branches" v-for="(branch, index) in branches" :key="index">
      <input type="radio"
             :id="branch"
             :value="branch"
             name="branch"
             v-model="currentBranch">
      <label :for="branch">{{ branch }}</label>
    </div>
    <p>vuejs/vue@{{ currentBranch }}</p>
    <ul>
      <li v-for="(record, ii) in commits" :key="ii">
        <a :href="record.html_url" target="_blank" class="commit">{{ record.sha.slice(0, 7) }}</a>
        - <span class="message">{{ record.commit.message | truncate }}</span><br>
        by <span class="author"><a :href="record.author.html_url"
                                   target="_blank">{{ record.commit.author.name }}</a></span>
        at <span class="date">{{ record.commit.author.date | formatDate }}</span>
      </li>
    </ul>
  </div>
</template>
<script>
import axios from 'axios'
var apiURL = 'https://api.github.com/repos/vuejs/vue/commits?access_token=e0c46e3543e2618353bc4d392e4f6016910622d0&per_page=3&sha='

export default {
  data: function () {
    return {
      branches: ['master', 'dev'],
      currentBranch: 'master',
      commits: null
    }
  },

  created: function () {
    this.fetchData()
  },

  watch: {
    currentBranch: 'fetchData'
  },

  filters: {
    truncate: function (v) {
      var newline = v.indexOf('\n')
      return newline > 0 ? v.slice(0, newline) : v
    },
    formatDate: function (v) {
      return v.replace(/T|Z/g, ' ')
    }
  },

  methods: {
    fetchData: function () {
      axios.get(apiURL + this.currentBranch).then(response => {
        this.commits = response.data
      })
    }
  }
}
</script>
<style>
  body {
    font-family: 'Helvetica', Arial, sans-serif;
  }

  a {
    text-decoration: none;
    color: #f66;
  }

  li {
    line-height: 1.5em;
    margin-bottom: 20px;
  }

  .branches {
    display: inline-block;
  }

  .author, .date {
    font-weight: bold;
  }
</style>
