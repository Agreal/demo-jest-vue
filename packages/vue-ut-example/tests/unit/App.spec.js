import {shallowMount} from '@vue/test-utils'
import axios from 'axios'
import flushPromises from 'flush-promises'
import App from '@/App.vue'

jest.mock('axios')

describe('App', function () {
  test('should render master commits list when mount App', async () => {
    axios.get.mockResolvedValue({data: stubMasterCommits()})

    const wrapper = shallowMount(App)
    await flushPromises()

    expect(wrapper.element).toMatchSnapshot()
    const $li = wrapper.findAll('li')
    expect($li.length).toBe(3);
    expect($li.at(0).find('a').text()).toBe('1234567');
  })

  test('should render dev commits list when checked dev branch', async () => {
    axios.get
      .mockResolvedValueOnce({data: stubMasterCommits()})
      .mockResolvedValueOnce({data: stubDevCommits()})

    const wrapper = shallowMount(App)
    await flushPromises()
    wrapper.setData({currentBranch: 'dev'})
    await flushPromises()

    expect(wrapper.element).toMatchSnapshot()
    const $li = wrapper.findAll('li')
    expect($li.length).toBe(3);
    expect($li.at(0).find('a').text()).toBe('1111111');
  })
});

function stubMasterCommits() {
  return [
    {
      html_url: 'html_url_0',
      sha: '1234567890',
      commit: {message: 'aaaaa', author: {name: 'aaa', date: '2017-10-13T03:07:14Z'}},
      author: {html_url: 'author_html_url'},
    },
    {
      html_url: 'html_url_1',
      sha: '1234567891',
      commit: {message: 'bbbbb', author: {name: 'bbb', date: '2017-10-13T03:07:14Z'}},
      author: {html_url: 'author_html_url'},
    },
    {
      html_url: 'html_url_2',
      sha: '1234567892',
      commit: {message: 'ccccc', author: {name: 'ccc', date: '2017-10-13T03:07:14Z'}},
      author: {html_url: 'author_html_url'},
    }
  ]
}

function stubDevCommits() {
  return [
    {
      html_url: 'html_url_0',
      sha: '1111111110',
      commit: {message: 'aaaaa', author: {name: 'aaa', date: '2017-10-13T03:07:14Z'}},
      author: {html_url: 'author_html_url'},
    },
    {
      html_url: 'html_url_1',
      sha: '1111111111',
      commit: {message: 'bbbbb', author: {name: 'bbb', date: '2017-10-13T03:07:14Z'}},
      author: {html_url: 'author_html_url'},
    },
    {
      html_url: 'html_url_2',
      sha: '1111111112',
      commit: {message: 'ccccc', author: {name: 'ccc', date: '2017-10-13T03:07:14Z'}},
      author: {html_url: 'author_html_url'},
    }
  ]
}
