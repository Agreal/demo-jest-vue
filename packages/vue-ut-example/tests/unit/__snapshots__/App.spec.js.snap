// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render dev commits list when checked dev branch 1`] = `
<div>
  <h1>
    Latest Vue.js Commits
  </h1>
   
  <div
    class="branches"
  >
    <input
      id="master"
      name="branch"
      type="radio"
      value="master"
    />
     
    <label
      for="master"
    >
      master
    </label>
  </div>
  <div
    class="branches"
  >
    <input
      id="dev"
      name="branch"
      type="radio"
      value="dev"
    />
     
    <label
      for="dev"
    >
      dev
    </label>
  </div>
   
  <p>
    vuejs/vue@dev
  </p>
   
  <ul>
    <li>
      <a
        class="commit"
        href="html_url_0"
        target="_blank"
      >
        1111111
      </a>
      
      - 
      <span
        class="message"
      >
        aaaaa
      </span>
      <br />
      
      by 
      <span
        class="author"
      >
        <a
          href="author_html_url"
          target="_blank"
        >
          aaa
        </a>
      </span>
      
      at 
      <span
        class="date"
      >
        2017-10-13 03:07:14 
      </span>
    </li>
    <li>
      <a
        class="commit"
        href="html_url_1"
        target="_blank"
      >
        1111111
      </a>
      
      - 
      <span
        class="message"
      >
        bbbbb
      </span>
      <br />
      
      by 
      <span
        class="author"
      >
        <a
          href="author_html_url"
          target="_blank"
        >
          bbb
        </a>
      </span>
      
      at 
      <span
        class="date"
      >
        2017-10-13 03:07:14 
      </span>
    </li>
    <li>
      <a
        class="commit"
        href="html_url_2"
        target="_blank"
      >
        1111111
      </a>
      
      - 
      <span
        class="message"
      >
        ccccc
      </span>
      <br />
      
      by 
      <span
        class="author"
      >
        <a
          href="author_html_url"
          target="_blank"
        >
          ccc
        </a>
      </span>
      
      at 
      <span
        class="date"
      >
        2017-10-13 03:07:14 
      </span>
    </li>
  </ul>
</div>
`;

exports[`App should render master commits list when mount App 1`] = `
<div>
  <h1>
    Latest Vue.js Commits
  </h1>
   
  <div
    class="branches"
  >
    <input
      id="master"
      name="branch"
      type="radio"
      value="master"
    />
     
    <label
      for="master"
    >
      master
    </label>
  </div>
  <div
    class="branches"
  >
    <input
      id="dev"
      name="branch"
      type="radio"
      value="dev"
    />
     
    <label
      for="dev"
    >
      dev
    </label>
  </div>
   
  <p>
    vuejs/vue@master
  </p>
   
  <ul>
    <li>
      <a
        class="commit"
        href="html_url_0"
        target="_blank"
      >
        1234567
      </a>
      
      - 
      <span
        class="message"
      >
        aaaaa
      </span>
      <br />
      
      by 
      <span
        class="author"
      >
        <a
          href="author_html_url"
          target="_blank"
        >
          aaa
        </a>
      </span>
      
      at 
      <span
        class="date"
      >
        2017-10-13 03:07:14 
      </span>
    </li>
    <li>
      <a
        class="commit"
        href="html_url_1"
        target="_blank"
      >
        1234567
      </a>
      
      - 
      <span
        class="message"
      >
        bbbbb
      </span>
      <br />
      
      by 
      <span
        class="author"
      >
        <a
          href="author_html_url"
          target="_blank"
        >
          bbb
        </a>
      </span>
      
      at 
      <span
        class="date"
      >
        2017-10-13 03:07:14 
      </span>
    </li>
    <li>
      <a
        class="commit"
        href="html_url_2"
        target="_blank"
      >
        1234567
      </a>
      
      - 
      <span
        class="message"
      >
        ccccc
      </span>
      <br />
      
      by 
      <span
        class="author"
      >
        <a
          href="author_html_url"
          target="_blank"
        >
          ccc
        </a>
      </span>
      
      at 
      <span
        class="date"
      >
        2017-10-13 03:07:14 
      </span>
    </li>
  </ul>
</div>
`;
